Update Connect to MetaMask with EIP-6963 TypeScript & React examples (#1247)
MetaMask/metamask-docs: wallet/how-to/connect/index.md

    }
}

const connectWithProvider = async (wallet: EIP6963AnnounceProviderEvent["detail"]) => {
    try {
        await wallet.provider
            .request({ method: "eth_requestAccounts" })
    } catch (error) {
        console.error("Failed to connect to provider:", error)
    }
}

export function listProviders(element: HTMLDivElement) {
    window.addEventListener("eip6963:announceProvider",
        (event: EIP6963AnnounceProviderEvent) => {
            const button = document.createElement("button")
        
            button.innerHTML = `
                <img src="${event.detail.info.icon}" alt="${event.detail.info.name}" />
                <div>${event.detail.info.name}</div>
            `
        
            button.onclick = () => connectWithProvider(event.detail)
            element.appendChild(button)
        }
    )

    window.dispatchEvent(new Event("eip6963:requestProvider"))
}
```

The `connectWithProvider` function connects the user to the selected provider using
[`eth_requestAccounts`](/wallet/reference/eth_requestaccounts).

The `wallet` object is passed as an argument to the function, indicating the argument type.

> The `listProviders` function uses a simplified approach (over mapping and joining an entire block of HTML).

Reviewer: Would it be better to say "...uses a simplified approach. Instead of mapping and joining an entire block of HTML, it directly passes the..."

It directly passes the `event.detail` object to the `connectWithProvider` function when a provider
is announced.

The `connectWithProvider` is called when a user selects the button.

`window.dispatchEvent` notifies event listeners and other parts of the dapp that a provider is being requested.

#### 5. View the project

Run the following command to view and test the Vite project in your browser:

```bash
npm run dev
```

#### Example

See the [Vanilla TypeScript example](https://github.com/MetaMask/vite-vanilla-ts-eip-6963) for more information.
You can clone the repository and run the example locally using `npm i && npm run dev`.

### React TypeScript

Follow these steps for creating a React TypeScript project to connect to MetaMask:

#### 1. Create a project

[Create a Vite project](https://v3.vitejs.dev/guide/#scaffolding-your-first-vite-project) using the

> template for Vanilla TypeScript or React TypeScript:

Reviewer: isn't this supposed to be "using the template for React TypeScript:"?

---
description: Connect to MetaMask via EIP-6963.
sidebar_position: 1
toc_max_heading_level: 4
---

# Connect to MetaMask

You can connect your dapp to users' MetaMask wallets by detecting MetaMask in their browsers and
connecting to their accounts.

> The recommended method for connecting to MetaMask or other browser wallets is using the wallet

Reviewer 2: I think this is in passive voice. Maybe this can be rewritten for active voice, such as:

We recommend using the wallet detection mechanism introduced by [EIP-6963](https://eips.ethereum.org/EIPS/eip-6963) for connecting to MetaMask and other browser wallets. This approach allows you to detect and connect to multiple wallets and without conflicts.

detection mechanism introduced by [EIP-6963](https://eips.ethereum.org/EIPS/eip-6963).
This method allows you to detect multiple wallets and connect to them without conflicts.

:::info
Learn more about EIP-6963 in [Wallet interoperability](../../concepts/wallet-interoperability.md).
:::

You can connect to MetaMask [using third-party libraries](#connect-to-metamask-using-third-party-libraries)
or [directly using Vite](#connect-to-metamask-directly-using-vite).

## Connect to MetaMask using third-party libraries

You can connect to MetaMask using the following third-party libraries that support EIP-6963:

- [Wagmi 2+](https://wagmi.sh)
- [Web3Modal 3+](https://docs.walletconnect.com/web3modal/about)
- [MIPD Store](https://github.com/wevm/mipd)
- [RainbowKit](https://www.rainbowkit.com)
- [Web3-Onboard](https://onboard.blocknative.com)

## Connect to MetaMask directly using Vite

To connect to MetaMask directly, we recommend implementing support for EIP-6963 using the

> [Vite](https://vitejs.dev/) build tool with [Vanilla TypeScript](#vanilla-typescript) or

Reviewer 2: I think "Vanilla Typescript" is not a proper noun, but installed "vanilla" is an adjective to describe the Typescript. I think maybe we should format instances as "vanilla Typescript".

        
            button.innerHTML = `
                <img src="${event.detail.info.icon}" alt="${event.detail.info.name}" />
                <div>${event.detail.info.name}</div>
            `
        
            // Call connectWithProvider when a user selects the button.
            button.onclick = () => connectWithProvider(event.detail)
            element.appendChild(button)
        }
    )

    // Notify event listeners and other parts of the dapp that a provider is requested.
    window.dispatchEvent(new Event("eip6963:requestProvider"))
}
```

The `connectWithProvider` function connects the user to the selected provider using
[`eth_requestAccounts`](/wallet/reference/eth_requestaccounts).
The `wallet` object is passed as an argument to the function, indicating the argument type.

The `listProviders` function uses a simplified approach.
Instead of mapping and joining an entire block of HTML, it directly passes the `event.detail` object
to the `connectWithProvider` function when a provider is announced.

#### 5. View the project

Run the following command to view and test the Vite project in your browser:

```bash
npm run dev
```

#### Example

> See the [Vanilla TypeScript example](https://github.com/MetaMask/vite-vanilla-ts-eip-6963) for more information.

Reviewer 2: ```suggestion
See the [vanilla TypeScript example](https://github.com/MetaMask/vite-vanilla-ts-eip-6963) for more information.
```

You can clone the repository and run the example locally using `npm i && npm run dev`.

### React TypeScript

Follow these steps for creating a React TypeScript project to connect to MetaMask:

#### 1. Create a project

[Create a Vite project](https://v3.vitejs.dev/guide/#scaffolding-your-first-vite-project) using the
template for React TypeScript:

```bash

> npm create vite@latest react-ts-6963 -- --template react-ts

Reviewer 2: Can this be more succinct or are the two sets of  `--`s needed to ? 

```suggestion
npm create vite@latest react-ts-6963 -- --template react-ts
```

Reviewer 3: Extra -- is needed for npm 7+, as per the linked Vite docs: https://v3.vitejs.dev/guide/#scaffolding-your-first-vite-project

sidebar_position: 1
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Connect to MetaMask

You can connect your dapp to users' MetaMask wallets by detecting MetaMask in their browsers and
connecting to their accounts.

> The best practice for detecting MetaMask or any other browser extension wallet (wallet provider) uses the wallet detection mechanism introduced by [EIP-6963](https://eips.ethereum.org/EIPS/eip-6963). If a user has multiple wallet browser extensions installed, you can detect multiple wallets and connect to each one without conflicts.

Reviewer 4: ```suggestion
The best practice for connecting to MetaMask or any other browser extension wallet (wallet provider) is to utilize the wallet detection mechanism introduced by [EIP-6963](https://eips.ethereum.org/EIPS/eip-6963). This method allows you to detect multiple wallets and connect to them in a way that avoids conflicts.
```

:::note

> [EIP-1193](https://eips.ethereum.org/EIPS/eip-1193) aimed to standardize the wallet interface, but conflicts emerged among implementations, causing race conditions. Wallets injecting providers clashed impacting user experience. This created UX issues for wallet discovery, onboarding, and connection, when multiple wallet extensions are enabled in the same browser. EIP-6963 solves this.

Reviewer 4: ```suggestion
[EIP-1193](https://eips.ethereum.org/EIPS/eip-1193) aimed to standardize the wallet interface, but suffers from unpredictable conflicts when multiple wallets are installed due to race conditions. Wallets injecting providers conflict with one another, which may cause UX issues across wallet discovery, onboarding, and connection, when multiple wallet extensions are enabled in the same browser. The pattern introduced by EIP-6963 solves these  issues.
```

- [Wagmi 2+](https://wagmi.sh)
- [Web3Modal 3+](https://docs.walletconnect.com/web3modal/about)
- [MIPD Store](https://github.com/wevm/mipd)
- [RainbowKit](https://www.rainbowkit.com)
- [Web3Onboard](https://onboard.blocknative.com)

## Connect to MetaMask with ViteJS

> For connecting to MetaMask, we suggest implementing support for EIP-6963 in JavaScript/TypeScript or React and use the

Reviewer 4: ```suggestion
For connecting to MetaMask, we suggest implementing support for EIP-6963 in JavaScript/TypeScript or React and using the
```
